crypto/http/http_client.c

```diff
@@ -0,0 +1,1199 @@
+                             headers, maxline, max_resp_len, timeout,
+                             expected_content_type, 1 /* expect_asn1 */))
+        != NULL)
+        resp = BIO_mem_d2i(mem, it);
+    BIO_free(mem);
+    return resp;
+}
+
+/* Post ASN.1-encoded request via HTTP to server return ASN.1 response */
+ASN1_VALUE *OSSL_HTTP_post_asn1(const char *server, const char *port,
+                                const char *path, int use_ssl,
+                                const char *proxy, const char *proxy_port,
+                                OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
+                                const STACK_OF(CONF_VALUE) *headers,
+                                const char *content_type,
+                                ASN1_VALUE *req, const ASN1_ITEM *req_it,
+                                int maxline, unsigned long max_resp_len,
+                                long timeout, const char *expected_ct,
+                                const ASN1_ITEM *rsp_it)
+{
+    BIO *req_mem;
+    BIO *res_mem;
+    ASN1_VALUE *resp = NULL;
+
+    if (req == NULL) {
+        HTTPerr(0, ERR_R_PASSED_NULL_PARAMETER);
+        return NULL;
+    }
+    /* remaining parameters are checked indirectly */
+
+    req_mem = HTTP_ASN1_item2BIO(req_it, req);
+    res_mem = OSSL_HTTP_transfer(server, port, path, use_ssl, proxy, proxy_port,
+                                 bio_update_fn, arg, headers, content_type,
+                                 req_mem /* may be NULL */, maxline,
+                                 max_resp_len, timeout,
+                                 expected_ct, 1 /* expect_asn1 */, NULL);
+    BIO_free(req_mem);
+    if (res_mem != NULL)
+        resp = BIO_mem_d2i(res_mem, rsp_it);
+    BIO_free(res_mem);
+    return resp;
+}
+
+/* BASE64 encoder used for encoding basic proxy authentication credentials */
+static char *base64encode(const void *buf, size_t len)
+{
+    int i;
+    size_t outl;
+    char *out;
+
+    /* Calculate size of encoded data */
+    outl = (len / 3);
+    if (len % 3 > 0)
+        outl++;
+    outl <<= 2;
+    out = OPENSSL_malloc(outl + 1);
+    if (out == NULL)
+        return 0;
+
+    i = EVP_EncodeBlock((unsigned char *)out, buf, len);
+    if (!ossl_assert((size_t)i <= outl)) {
+        OPENSSL_free(out);
+        return NULL;
+    }
+    if (i < 0)
+        *out = '\0';
+    return out;
+}
+
+/*
+ * Promote the given connection BIO using the CONNECT method for a TLS proxy.
+ * This is typically called by an app, so bio_err and prog are used unless NULL
+ * to print additional diagnostic information in a user-oriented way.
+ */
+int OSSL_HTTP_proxy_connect(BIO *bio, const char *server, const char *port,
+                            const char *proxyuser, const char *proxypass,
+                            long timeout, BIO *bio_err, const char *prog)
+{
+# undef BUF_SIZE
+# define BUF_SIZE (8 * 1024)
+    char *mbuf = OPENSSL_malloc(BUF_SIZE);
```

Comment:
I don't know what the reasoning was since I basically just moved this function.
Yet in my view allocating 8k on the stack is a bit much. 
I know that in other situations (at most) 4k are allocated.